testing/opae-c/test_init_c.cpp

```diff
@@ -166,3 +172,245 @@ TEST(init, log_file) {
   unlink("opae_log.log");
 }
 
+/**
+ * @test       find_ase_cfg
+ *
+ * @brief      When WITH_ASE is specified, opae_ase.cfg will
+ *             be searched from OPAE source directory, OPAE
+ *             installation directory or home/system config directory.
+ *             
+ */
+TEST(init, find_ase_cfg) {
+  char *cfg_path = nullptr;
+
+  ASSERT_EQ(0, putenv((char*)"WITH_ASE=1"));
+  cfg_path = find_ase_cfg();
+  EXPECT_NE(cfg_path, nullptr);
+
+  EXPECT_EQ(0, unsetenv("WITH_ASE"));
+}
+
+const char *dummy_cfg = R"plug(
+{
+    "configurations": {
+        "dummy": {
+            "configuration": {
+                "key1": "hello",
+                "key2": "plugin",
+                "fake_tokens": 99
+            },
+            "enabled": true,
+            "plugin": "libdummy_plugin.so"
+        }
+    },
+    "plugins": [
+        "dummy"
+    ]
+}
+)plug";
+
+class init_ase_cfg_p : public ::testing::TestWithParam<const char*> {
+ protected:
+  init_ase_cfg_p() : buffer_ {0} {}
+
+  virtual void SetUp() override {
+    // let's rename the opae_ase.cfg in OPAE_ASE_CFG_SRC_PATH and OPAE_ASE_CFG_INST_PATH
+
+    // copy it to a temporary buffer that we can use dirname with
+    std::string src_cfg_path = (OPAE_ASE_CFG_SRC_PATH? OPAE_ASE_CFG_SRC_PATH : "");
+    std::copy(src_cfg_path.begin(), src_cfg_path.end(), &buffer_[0]);
+    char *src_cfg_dir = dirname(buffer_);
+    char *src_cfg_file = basename(buffer_);
```

Comment:
Will do null pointer checking.

Reply:
Right.

Reply:
Null checks won't help.  man dirname says that these functions always return a pointer to a NULL-terminated string.  You will need to see what the invalid values are and use strcmp.

Reply:
Below are the definitions in config.h.in: 
#define OPAE_ASE_CFG_SRC_PATH "@CMAKE_CURRENT_SOURCE_DIR@/libopae/plugins/ase/opae_ase.cfg"
#define OPAE_ASE_CFG_INST_PATH "@CMAKE_INSTALL_PREFIX@/share/opae/ase/opae_ase.cfg"

It will always return some directory name by definition. Do you think we still need to check?

Reply:
It's probably not critical; however please remove the nullptr checks for clarity.